### PR TITLE
feat(julia): support the julia language

### PR DESCRIPTION
**Pull request by brandonspark**
image

## Commentary:
It's probably easier to read the file directly than it is to try and look at the diff.

The grammar changes included in `semgrep-julia` were so drastic from the last time I looked at it, that I had to take the previous `Parse_julia_tree_sitter.ml` I wrote and basically scrap it, because so much of it was now different. As such, perhaps it's best to look at this file with a clear eye.

I eliminated pretty much every single TODO, in favor of making sacrifices and tradeoffs that would allow us to parse for now, if at the cost of some semantic information. In particular, I've made liberal use of `Other`, as Julia's implementation means that quite often, things can appear in places that we do not expect. 

In particular, statements and expressions have really no meaning, macro-expressions can appear _everywhere_ an identifier can (making things like macro definitions, imports, etc more difficult), and it's annoying that things which should be patterns are represented in the grammar as just plain expressions. For the latter, I've just injected them into an `Other` for now. Comments have been left, hopefully clearing up the places that I have had to make these tradeoffs, other than just everywhere an `Other` appears.

## Test plan:
`./run-lang julia`

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: parsing-stats/lang/julia/projects.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#
+# Top julia projects from GitHub, sorted by stars.
+# Created by /Users/brandonspark/semgrep/semgrep-core/src/ocaml-tree-sitter-core/scripts/most-starred-for-language.
+#
+https://github.com/JuliaLang/julia
```
Comment (aryx) [inline parsing-stats/lang/julia/projects.txt]: This we could potentially comment because it probably contact lots of test cases exercising edge cases
in the grammar that are not really representative of typical julia programs (we do that for other languages).

File: semgrep-core/src/parsing/tree_sitter/Parse_julia_tree_sitter.ml [hunk 2]
```diff
@@ -1,47 +1,64 @@
 (* Yoann Padioleau
```
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_julia_tree_sitter.ml]: At this point you can put your name here.

File: semgrep-core/src/parsing/tree_sitter/Parse_julia_tree_sitter.ml [hunk 3]
```diff
@@ -1,47 +1,64 @@
...
+   * directly to AST_generic.ml
+   *
+*)
 
 (*****************************************************************************)
 (* Helpers *)
 (*****************************************************************************)
 
+(* Disable warnings against unused variables *)
+[@@@warning "-26-27"]
```
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_julia_tree_sitter.ml]: This we want to remove ASAP.

File: semgrep-core/src/parsing/tree_sitter/Parse_julia_tree_sitter.ml [hunk 4]
```diff
@@ -1,47 +1,64 @@
...
+
+(* Disable warning against unused 'rec' *)
+[@@@warning "-39"]
+
 type env = unit H.env
 
+let ( let* ) = Option.bind
```
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_julia_tree_sitter.ml]: I think if you open Common you get this let* too.

**Review by aryx — APPROVED** [review]
Great job!
Now we probably need to start writing alpha tests for alpha
to check if everything works fine. Probably need to extend the ocaml-tree-sitter-semgrep/lang/..../semgrep-julia/grammar.js and add more semgrep extensions (..., metavars, etc.)

**Comment by aryx** [comment]
Regarding expressions as patterns, this is a recurring problem with other languages (e.g., Python). AST_generic_helpers.ml provide some helpers such as expr_to_pattern(), but it's quite limited.
AST_generic.ml has also a few useful helpers at the end of the file (e.g., one to build opcall).

**Review by aryx — APPROVED** [review]
